### PR TITLE
hltGetConfiguration updates to deal with FinalPaths and new OutputModule

### PR DESCRIPTION
**Pull request by Sam-Harper**
#### PR description:

This PR updates hltGetConfiguration and sub table creation to handle the new HLT configs which will use FinalPaths. It also allows sub table creation to use the test converter and dev database.   The generated HLT output paths for offline use are also updated to be FinalPaths.

Finally it adapts hltGetConfiguration to handle the new OutputModule GlobalEvtOutputModule. 

No changes are expected when retrieving existing configs except the generated output paths will be FinalPaths.

#### PR validation:

subtable creation with the new converter works as expected

```
hltGetConfiguration /dev/CMSSW_12_3_0/GRun --offline --full --mc --globaltag auto:phase1_2021_realistic 
```
gives identical results before and after this fix.  With `--output full` , the difference is the output path is now a FinalPath

```
hltGetConfiguration v3-test/dev:/users/sharper/2022/CMSSW_12_3_0_V27_DatasetPathV2/GRun --offline --full --mc --globaltag auto:phase1_2021_realistic 
```

produces a runable config.

### CODE REVIEW

File: HLTrigger/Configuration/tables/subtables.sh [hunk 1]
```diff
@@ -110,6 +110,20 @@ function loadConfiguration() {
...
+      DBNAME="cms_hlt.cern.ch"
+      DBUSER="cms_hlt_gdrdev_w"
+      PWHASH="0196d34dd35b04c0f3597dc89fbbe6e2"
+      ;;
+      "v3/dev")
+      # v3 dev
+      DBHOST="cmsr1-s.cern.ch"
+      DBNAME="cms_hlt.cern.ch"
+      DBUSER="cms_hlt_gdrdev_w"
+      PWHASH="0196d34dd35b04c0f3597dc89fbbe6e2"
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/tables/subtables.sh:125]: It seems "v3-test/dev" and "v3/dev" have the very same parameters, shouldn't there be a difference?

Comment (Sam-Harper) [reply HLTrigger/Configuration/tables/subtables.sh:125]: dev is the database and thats what the parameters reflect.  v3 and v3-test are the converters to use. Honestly should have been rewritten to remove the converter from the string but meh...

Comment (Martin-Grunewald) [reply HLTrigger/Configuration/tables/subtables.sh:125]: OK, thanks for the clarification

**Comment by Martin-Grunewald** [comment]
please test

**Comment by missirol** [comment]
Looks like some type of fix is also needed in
https://github.com/cms-sw/cmssw/blob/ec6f44d67fa797c448798442f99ed567a9f99680/HLTrigger/Configuration/python/Tools/confdb.py#L285-L290

As an example, the following does not return a valid python config:
```bash
# using a GRun subtable created from a master table converted to Dataset paths
hltGetConfiguration --cff --data v3-test/dev:/users/missirol/test/taskify/test01/GRun/V2 --type GRun > tmp_GRun.cff
python3 tmp_GRun.cff
```

**Comment by Sam-Harper** [comment]
So many thanks Marino for this. After our mattermost conversation, I pushed a (inelegant) fix which we had discussed. 

And also just for others here, in our chat Marino pointed out something nasty with cmsDriver.py, ie it needs to use FinalPath output modules and this is going to be a lot of fun.... 

so to get working in cmsDriver, its  a bit of a bigger job and will need a bit of testing

**Comment by Martin-Grunewald** [comment]
> So many thanks Marino for this. After our mattermost conversation, I pushed a (inelegant) fix which we had discussed.
> 
> And also just for others here, in our chat Marino pointed out something nasty with cmsDriver.py, ie it needs to use FinalPath output modules and this is going to be a lot of fun....
> 
> so to get working in cmsDriver, its a bit of a bigger job and will need a bit of testing

What is the issue here? cmsDriver uses the HLT cff which does not contain output modules/output  paths.

**Comment by missirol** [comment]
While I was testing this PR and the ConfDB updates, I went one step further and used Tasks (converting a menu to Tasks with the "Taskify" button). For that menu, the TSG tests where we run HLT+RECO together did not pass (the TSG tests running the HLT step alone passed); the problem comes from the fact that cmsDriver adds EndPaths with OutputModules, and this can trigger the premature (unscheduled) execution of HLT modules (i.e. the original problem).

But this is a potential problem with Tasks, not with this PR per se. Still, I would like to do some more checks before merging this PR as it is.

**Comment by Martin-Grunewald** [comment]
EndPaths with what OutputModules, or more precisely, EventContent?

**Comment by missirol** [comment]
These are three examples of errors I have seen (I think the other errors are repetitions of these 3).

 1. `RelVal_HLT_GRun_MC.log`:
  ```
----- Begin Fatal Exception 04-Feb-2022 22:15:31 CET-----------------------                                                                                                                        
An exception of category 'ProductNotFound' occurred while                                                                                                                                          
   [0] Processing  Event run: 1 lumi: 2 event: 54 stream: 1                                                                                                                                        
   [1] Running path 'FEVTDEBUGHLToutput_step'                                                                                                                                                      
   [2] Prefetching for module PoolOutputModule/'FEVTDEBUGHLToutput'                                                                                                                                
   [3] Calling method for module HLTCaloJetCollectionProducer/'hltCaloJetsSix30ForHt'                                                                                                              
Exception Message:                                                                                                                                                                                 
Principal::getByToken: Found zero products matching all criteria                                                                                                                                   
Looking for type: trigger::TriggerFilterObjectWithRefs                                                                                                                                             
Looking for module label: hltCaloJetFilterSixC30                                                                                                                                                   
Looking for productInstanceName:                                                                                                                                                                   
                                                                                                                                                                                                   
   Additional Info:                                                                                                                                                                                
      [a] If you wish to continue processing events after a ProductNotFound exception,                                                                                                             
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.                                                                                             
                                                                                                                                                                                                   
----- End Fatal Exception -------------------------------------------------                                                                                                                        
  ``` 
 2. `RelVal_HLT_Reco_GRun_MC.log`:
  ```
----- Begin Fatal Exception 04-Feb-2022 22:50:29 CET-----------------------                                                                                                                        
An exception of category 'ProductNotFound' occurred while                                                                                                                                          
   [0] Processing  Event run: 1 lumi: 2 event: 57 stream: 0                                                                                                                                        
   [1] Running path 'RAWoutput_step'                                                                                                                                                               
   [2] Prefetching for module PoolOutputModule/'RAWoutput'                                                                                                                                         
   [3] Prefetching for module TriggerSummaryProducerAOD/'hltTriggerSummaryAOD'                                                                                                                     
   [4] Calling method for module HLTPFJetCollectionProducer/'hltPFJetForDBtagAK8'                                                                                                                  
Exception Message:                                                                                                                                                                                 
Principal::getByToken: Found zero products matching all criteria                                                                                                                                   
Looking for type: trigger::TriggerFilterObjectWithRefs                                                                                                                                             
Looking for module label: hltPFJetForDBtagSelectorAK8                                                                                                                                              
Looking for productInstanceName:                                                                                                                                                                   
                                                                                                                                                                                                   
   Additional Info:                                                                                                                                                                                
      [a] If you wish to continue processing events after a ProductNotFound exception,                                                                                                             
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.                                                                                             
                                                                                                                                                                                                   
----- End Fatal Exception -------------------------------------------------                                                                                                                        
  ```

 3. `RelVal_DigiL1RawHLT_GRun_MC.log`:
  ```
----- Begin Fatal Exception 04-Feb-2022 22:52:17 CET-----------------------                                                                                                                        
An exception of category 'ProductNotFound' occurred while                                                                                                                                          
   [0] Processing  Event run: 1 lumi: 2 event: 55 stream: 0                                                                                                                                        
   [1] Running path 'FEVTDEBUGHLToutput_step'                                                                                                                                                      
   [2] Prefetching for module PoolOutputModule/'FEVTDEBUGHLToutput'                                                                                                                                
   [3] Calling method for module L1THLTTauMatching/'hltHpsL1JetsHLTDoublePFTauTrackPt1TightChargedIsolationMatch'                                                                                  
Exception Message:                                                                                                                                                                                 
Principal::getByToken: Found zero products matching all criteria                                                                                                                                   
Looking for type: std::vector<reco::PFTau>                                                                                                                                                         
Looking for module label: hltHpsSelectedPFTausTrackPt1TightChargedIsolation                                                                                                                        
Looking for productInstanceName:                                                                                                                                                                   
                                                                                                                                                                                                   
   Additional Info:                                                                                                                                                                                
      [a] If you wish to continue processing events after a ProductNotFound exception,                                                                                                             
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.                                                                                             
                                                                                                                                                                                                   
----- End Fatal Exception -------------------------------------------------                                                                                                                        
  ```

Example 1. shows that my previous statement is inaccurate

> the TSG tests where we run HLT+RECO together did not pass (the TSG tests running the HLT step alone passed)

because there is also a failure when HLT runs alone, but cmsDriver adds the EndPath `FEVTDEBUGHLToutput_step`.

**Comment by Sam-Harper** [comment]
Yes we just need to swap these endpaths to final paths in cmsDriver

However if I understand it the issue is that

HLT must not have unscheduled execution on its output modules
RECO must have unscheduled execution on its output modules (and I'm using RECO to mean offline, eg MiniAOD production)

so we have a problem if we ever want to run HLT and RECO in the same job (which I dont think is possible anyways or at least wasnt possible for along time). We also have to be careful of what else we are running in a HLT job. 

All this feels solvable though although maybe not in this PR which I would say is beyond its scope.

**Comment by Martin-Grunewald** [comment]
@Sam-Harper 
Running HLT+RECO in the same cmsRun was/is always (technically) possible, it is just that the physics output was never validated (eg, potential problems with RECO ES modules overwriting HLT ES modules etc.)

**Comment by Sam-Harper** [comment]
Ah thanks for the clarification Martin!

**Comment by Martin-Grunewald** [comment]
Does it mean that RECO OMs use HLT event contents or event contents pertaining to HLT modules?

**Comment by Sam-Harper** [comment]
> Does it mean that RECO OMs use HLT event contents or event contents pertaining to HLT modules?

Isnt there just one outmodule per job offline (modulo DQM)? And RECO and HLT have conflicting needs on whether it triggers unscheduled or not. 

So running RECO and HLT in the same job (or HLT and any other step but I suspect the problem is severe for others steps) is a problem right?

**Comment by Martin-Grunewald** [comment]
What I mean is that (unscheduled) execution of a module is triggered by the keep statement in the OM, so this OM must refer to HLT modules, no? Or are simply all EDProducers triggered to run regardless?

**Comment by Sam-Harper** [comment]
yes the unscheduled execution of a module is triggered by the keep statement in the OM and yes offline sometimes writes out HLT products, particularly in FEVTDEBUG.  Therefore that OM must be on a finalpath

The issue is that perhaps the HLT is run with something else which needs the unscheduled execution and thus that OM can not be on a finalpath.  Otherwise we would just have cmsDriver swap the module output path to a finalpath and solve the issue trivaly

**Comment by Martin-Grunewald** [comment]
+1
Solving (un)scheduled/TASK problem relegated to a subsequent PR, as this is fine for current standard (non-TASK) HLT menus.

**Comment by perrotta** [comment]
+1
- Approved by HLT

**Pull request merged**